### PR TITLE
Update logits processor with tensor caching

### PR DESCRIPTION
**Pull request by lynkz-matt-psaltis**
Re-adds caching for tensor lookup on top of latest vllm and outlines changes.

### CODE REVIEW

**Review by lapp0 — CHANGES_REQUESTED** [review]
`outlines.processors` are compatible with vllm with no changes: https://github.com/outlines-dev/outlines/blob/main/outlines/processors/structured.py

We have [benchmark tests for torch cpu / gpu](https://github.com/outlines-dev/outlines/blob/main/benchmarks/bench_processors.py#L26-L35) use of these logits processors and [integration tests with vLLM](https://github.com/outlines-dev/outlines/blob/main/tests/generate/test_generate.py#L72-L74) to ensure performance and compatibility.

Is there any reason to not just have this module run `from outlines.processors import RegexLogitsProcessor, JSONLogitsProcessor`?

**Comment by lynkz-matt-psaltis** [comment]
> `outlines.processors` are compatible with vllm with no changes: https://github.com/outlines-dev/outlines/blob/main/outlines/processors/structured.py
> 
> We have [benchmark tests for torch cpu / gpu](https://github.com/outlines-dev/outlines/blob/main/benchmarks/bench_processors.py#L26-L35) use of these logits processors and [integration tests with vLLM](https://github.com/outlines-dev/outlines/blob/main/tests/generate/test_generate.py#L72-L74) to ensure performance and compatibility.
> 
> Is there any reason to not just have this module run `from outlines.processors import RegexLogitsProcessor, JSONLogitsProcessor`?

I like where your head's at :)

I don't see any reason to maintain the additional abstraction any more.
I'll do up a separate PR based on that and make sure we've got a perf test in for these scenarios specifically.
Is there a slash command or similar that will run the benchmarks on PRs?

**Comment by lapp0** [comment]
> I like where your head's at :)
> 
> I don't see any reason to maintain the additional abstraction any more. I'll do up a separate PR based on that and make sure we've got a perf test in for these scenarios specifically. Is there a slash command or similar that will run the benchmarks on PRs?

That's great to hear, thanks so much for making this change!

Here is documentation on using `asv` within outlines https://github.com/outlines-dev/outlines/blob/main/docs/community/contribute.md#benchmarking

Here's how we test torch logits in with `outlines.processors` 
- module https://github.com/outlines-dev/outlines/blob/main/benchmarks/bench_processors.py
- command (in outlines repo directory) `asv run --config benchmarks/asv.conf.json -b bench_processors.LogitsProcessorStructuredBenchmark.time_structured_generation`

Please let me know if you have any other questions or if there's any other way I can help!
